src/libmongoc/src/mongoc/mongoc-client.c

```diff
@@ -1084,54 +1071,48 @@ mongoc_client_set_ssl_opts (mongoc_client_t *client,
 #endif
 
 
-/*
- *--------------------------------------------------------------------------
- *
- * mongoc_client_new_from_uri --
- *
- *       Create a new mongoc_client_t for a mongoc_uri_t.
- *
- * Returns:
- *       A newly allocated mongoc_client_t.
- *
- * Side effects:
- *       None.
- *
- *--------------------------------------------------------------------------
- */
-
 mongoc_client_t *
 mongoc_client_new_from_uri (const mongoc_uri_t *uri)
+{
+   mongoc_client_t *client;
+   bson_error_t error = {0};
+
+   if (!(client = mongoc_client_new_from_uri_with_error (uri, &error))) {
+      MONGOC_ERROR ("%s", error.message);
+   }
+
+   return client;
+}
+
+
+mongoc_client_t *
+mongoc_client_new_from_uri_with_error (const mongoc_uri_t *uri,
+                                       bson_error_t *error)
 {
    mongoc_topology_t *topology;
 
    topology = mongoc_topology_new (uri, true);
 
+   if (!topology->valid) {
+      if (error) {
+         memcpy (error, &topology->scanner->error, sizeof (bson_error_t));
+      }
+
+      mongoc_topology_destroy (topology);
+
+      return NULL;
+   }
+
    /* topology->uri may be different from uri: if this is a mongodb+srv:// URI
     * then mongoc_topology_new has fetched SRV and TXT records and updated its
     * uri from them.
     */
-   return _mongoc_client_new_from_uri (topology);
+   return _mongoc_client_new_from_topology (topology);
```

Comment:
I suggest storing the return value and freeing `topology` if `_mongoc_client_new_from_topology` returns `NULL`. Otherwise this may leak `topology`.

Reply:
Thanks. The two cases where `NULL` is returned would be from an invalid topology (addressed above) and TLS not being supported (not yet addressed). Will fix this now.